cupy/random/generator.py

```diff
@@ -341,29 +349,30 @@ def multivariate_normal(self, mean, cov, size=None, check_valid='ignore',
             raise ValueError('mean and cov must have same length')
         shape += (len(mean),)
 
-        x = self.standard_normal(size=shape, dtype=dtype)
-
-        u, s, v = cupy.linalg.svd(cov)
-
         if check_valid != 'ignore':
             if check_valid != 'warn' and check_valid != 'raise':
                 raise ValueError(
-                    'check_valid must equal \'warn\', \'raise\', or '
-                    '\'ignore\'')
-
-            a = cupy.dot(v.T * s, v)
-            b = cov
-            psd = cupy.all(cupy.abs(a-b) <= tol*(1+cupy.abs(b)))
+                    "check_valid must equal 'warn', 'raise', or 'ignore'")
+            else:
+                from cupy.linalg import eigh
+                (s, u) = eigh(cov)
+                psd = not cupy.any(s < -tol)
             if not psd:
                 if check_valid == 'warn':
-                    warnings.warn(
-                        'covariance is not symmetric positive-semidefinite.',
-                        RuntimeWarning)
+                    warnings.warn("covariance is not positive-semidefinite.",
+                                  RuntimeWarning)
                 else:
-                    raise ValueError(
-                        'covariance is not symmetric positive-semidefinite.')
+                    raise ValueError("covariance is not " +
+                                     "positive-semidefinite.")
 
-        x = cupy.dot(x, cupy.sqrt(s)[:, None] * v)
+        x = self.standard_normal(size=shape, dtype=dtype)
+        seterr(linalg='raise')
```

Comment:
This will force device synchronization every time,
how about `seterr(linalg=check_valid)` I think it is not as precise as numpy but is ok just to document this.
@niboshi what are your thoughts?

Reply:
Also, please use `cupyx.errstate` since its a context manager and will restore the configuration to what it was before.
`with cupyx.errstate(linalg=check_valid`

Reply:
Ok, I use the context manager now. I think that we want `linalg='raise'` instead of `linalg=check_state` though; otherwise the cholesky decomposition example that you gave with `a=cupy.arange(9).reshape(3,3)` will not raise an exception to mirror numpy behavior on equivalent input. I checked, and `linalg` does revert back to previous state after the function call. Being able to use check tolerance is probably unnecessary besides for not breaking code, but maybe it still has utility in some cases? If `check_valid` is used, it certainly will verify that the covariance matrix is valid by an independent call to `einh`, so it works as advertised... I just can't think of a case where cholesky doesn't fail but the covariance matrix is invalid, but perhaps there is a corner case that I don't know of.

Reply:
I guess the issue is that `cupyx.errstate` only has `ignore` and `raise` and there is no `warn` implemented; so do you want to be conservative and `raise` more than numpy does? Or ignore and **not** have run time warnings for invalid input, even though numpy will warn regardless of `check_valid`? I think raising is reasonable, but I'm curious what behavior you want to target...

Reply:
Thanks for the changes!
Yes, I understand that. However, in the current state, this function will raise an exception even if `check_valid` is set to ignore.
The problem here is that `errstate(linalg='raise')` forces device synchronization thus degrading performance, so we should avoid it in the case of `check_valid='ignore'`.

I am in favor of emulating the warn behavior by using raise and catching the exception.
I am sorry that my explanations are a bit confusing and you have to keep working on it, but we are close to finish it 😅

Reply:
@emcastillo How is the behavior now? Does it match what you would like?

Reply:
Yes, it looks better now!
Thanks a lot for the changes